kolena/io.py

```diff
@@ -29,33 +30,29 @@
 from kolena._utils.datatypes import DataObject
 
 
-def _serialize_dataobject(x: Any) -> Any:
-    if isinstance(x, list):
-        return [item._to_dict() if isinstance(item, DataObject) else item for item in x]
-
-    return x._to_dict() if isinstance(x, DataObject) else x
+class DataObjectJSONEncoder(json.JSONEncoder):
+    def default(self, o: Any) -> Dict:
+        if isinstance(o, DataObject):
+            return o._to_dict()
+        return super().default(o)
 
 
 def _deserialize_dataobject(x: Any) -> Any:
     if isinstance(x, list):
```

Comment:
Unfortunately, there isn't a list hook for JSONDecoder https://docs.python.org/3/library/json.html#encoders-and-decoders, so I opted to extend this solution.

Reply:
This is the code I tested
```
class CsutomDecoder(json.JSONDecoder):
    def __init__(self, *args, **kwargs):
        json.JSONDecoder.__init__(self, object_hook=self.object_hook, *args, **kwargs)

    def object_hook(self, dct):
        if 'data_type' in dct:
            return Label(label=str(dct["data_type"]))
        return dct

print(
    json.loads(
        '[{"data_type": "aaa"}, {"x": 12, "y": 13, "c": [{"data_type": 123}, {"data_type": 456}]}, "abc", 12.4]',
        cls=CsutomDecoder
    )
)
print(
    json.loads(
        '{"x": 12, "y": 13, "c": [{"data_type": 123}, {"data_type": 456}, {"a": 12, "b": {"data_type": "sdf"}}]}',
        cls=CsutomDecoder
    )
)
```
and output is
```
[Label(label='aaa'), {'x': 12, 'y': 13, 'c': [Label(label='123'), Label(label='456')]}, 'abc', 12.4]
{'x': 12, 'y': 13, 'c': [Label(label='123'), Label(label='456'), {'a': 12, 'b': Label(label='sdf')}]}
```

Did I miss anything?

Reply:
I had given up too soon.. I was missing passing the object hook to the base decoders `__init__`. Thanks for checking this out!

Unfortunately we can't get rid of `_deserialize_dataobject` because https://pandas.pydata.org/pandas-docs/version/1.5/reference/api/pandas.read_json.html does not expose it's json decoder https://github.com/kolenaIO/kolena/blob/trunk/kolena/io.py#L106-L107